### DISCUSSION TITLE
How to read baseclass attribute

### DISCUSSION THREAD

**Discussion by snehabihani** [Washi1337/AsmResolver#419] [Q&A]
```  
[AttributeUsage(validOn: AttributeTargets.Field, AllowMultiple = true, Inherited = true)]
public class OutputExecutionPortAttribute : BasePortAttribute
{
}
public class BasePortAttribute : Attribute
{
        public string Description { get; set; }
}
    
[OutputExecutionPort] public Func<Task> OnNext;
```  
I want to get attribute "OutputExecutionPort" properties including base class.
Please help

**Comment by Washi1337** [comment]
If I understood you correctly, you are looking to find all the properties of an attribute attached to a field, including the properties defined in any of its base classes.

Assuming you have your field stored in a `FieldDefinition field`, you can get its custom attributes and its corresponding type definition using something like the following:
```csharp
var attribute = field.CustomAttributes.First(x => x.Constructor?.DeclaringType.Name == "OutputExecutionPortAttribute");
var type = attribute.Constructor.DeclaringType.Resolve();
```

Then, it is a matter of iterating all its properties, while also iterating all base types of the type:
```csharp
var properties = new List<PropertyDefinition>();
while (type is not null)
{
   properties.AddRange(type.Properties);
   type = type.BaseType?.Resolve();
}
```

**Comment by snehabihani** [comment]
yes. Thanks for quick reply. One more last query in below I need all
ConfigurationParameter when I start with TestNode Class

TestNode : PlatformVisualNode<TestNodeConfiguration>
{}
public class TestNodeConfiguration : BaseNodeConfiguration
{
    [ConfigurationParameter]
    public string text { get; set; }
}

 public class BaseNodeConfiguration : BaseConfiguration
    {
 [ConfigurationParameter]
        public bool IsProvider { get; set; }
}
 public class BaseNodeConfiguration :Interface
{ [ConfigurationParameter]
        public bool IsProvider1 { get; set; }
}

**Reply by Washi1337** [reply]
It would be a similar approach, where you'd traverse all types, their properties and read their custom attributes. You can read more about traversing the member tree on [this page](https://asmresolver.readthedocs.io/en/latest/dotnet/member-tree.html). Also, `PlatformVisualNode<TestNodeConfiguration>` is going to be a `TypeSpecification` with a `TypeSignature` in it, you may also want to have a look at the [docs on type signatures](https://asmresolver.readthedocs.io/en/latest/dotnet/type-signatures.html) as well :).

**Comment by snehabihani** [comment]
Can you please help with typesignature ? Today i was struggling to get that

On Wednesday, March 8, 2023, Washi ***@***.***> wrote:

> It would be a similar approach, where you'd traverse all types, their
> properties and read their custom attributes. You can read more about
> traversing the member tree on this page
> <https://asmresolver.readthedocs.io/en/latest/dotnet/member-tree.html>.
> Also, PlatformVisualNode<TestNodeConfiguration> is going to be a
> TypeSpecification with a TypeSignature in it, you may also want to have a
> look at the docs on type signatures
> <https://asmresolver.readthedocs.io/en/latest/dotnet/type-signatures.html>
> as well :).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Washi1337/AsmResolver/discussions/419#discussioncomment-5243952>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM4MS66MQCXNUA5ZZWATHOTW3CY23ANCNFSM6AAAAAAVTVXX4E>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Regards,
Sneha Bihani
Mob:9960298364

**Comment by snehabihani** [comment]
Please help

On Wednesday, March 8, 2023, Sneha Bihani ***@***.***> wrote:

> Can you please help with typesignature ? Today i was struggling to get
> that
>
> On Wednesday, March 8, 2023, Washi ***@***.***> wrote:
>
>> It would be a similar approach, where you'd traverse all types, their
>> properties and read their custom attributes. You can read more about
>> traversing the member tree on this page
>> <https://asmresolver.readthedocs.io/en/latest/dotnet/member-tree.html>.
>> Also, PlatformVisualNode<TestNodeConfiguration> is going to be a
>> TypeSpecification with a TypeSignature in it, you may also want to have
>> a look at the docs on type signatures
>> <https://asmresolver.readthedocs.io/en/latest/dotnet/type-signatures.html>
>> as well :).
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Washi1337/AsmResolver/discussions/419#discussioncomment-5243952>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AM4MS66MQCXNUA5ZZWATHOTW3CY23ANCNFSM6AAAAAAVTVXX4E>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Regards,
> Sneha Bihani
> Mob:9960298364
>
>

-- 
Regards,
Sneha Bihani
Mob:9960298364

**Comment by Washi1337** [comment]
You have to be a bit more specific with what you'd need help with :sweat_smile:. What part do you struggle with? 

`PlatformVisualNode<TestNodeConfiguration>` is going to be a `GenericInstanceTypeSignature` wrapped in a `TypeSpecification`. You'd have to check whether the type's base type is such a type specification, and act accordingly.

**Comment by snehabihani** [comment]
How to get type signature and tyere attributes

On Thursday, March 9, 2023, Washi ***@***.***> wrote:

> You have to be a bit more specific with what you'd need help with 😅.
> What part do you struggle with?
>
> PlatformVisualNode<TestNodeConfiguration> is going to be a
> GenericInstanceTypeSignature wrapped in a TypeSpecification. You'd have
> to check whether the type's base type is such a type specification, and act
> accordingly.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Washi1337/AsmResolver/discussions/419#discussioncomment-5247272>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM4MS647BFWE44NCY6LEHFDW3EFWVANCNFSM6AAAAAAVTVXX4E>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Regards,
Sneha Bihani
Mob:9960298364

**Comment by snehabihani** [comment]
var typeSignature = basetype.ToTypeSignature(); if
(typeSignature.ElementType == ElementType.GenericInst) { }
How to get list of arguments from here
Regards,
Sneha Bihani
Mob:9960298364

**Comment by snehabihani** [comment]
I got it. Thanks a lot for quick response.

**Comment by snehabihani** [comment]
I have struct defined in another assembly not referenced in my project but
I want to create instance for that how we can do?

 Activator.CreateInstance(System.Type.GetType(prop.Type.FullName)
Microsoft.Maui.Gridlenth but it is not referenced.

**Reply by Washi1337** [reply]
AsmResolver is a static analysis library. It cannot create new instances of a type.
